python/symbolica_community/tensors.pyi

```diff
@@ -0,0 +1,307 @@
+
+class CompiledTensorEvaluator:
+    r"""
+    A compiled and optimized evaluator for tensors.
+    """
+
+    def evaluate(self,inputs:Sequence[Sequence[float]])-> List[Tensor]:
+        """Evaluate the expression for multiple inputs and return the result."""
+
+    def evaluate_complex(self, inputs: Sequence[Sequence[complex]]) -> List[Tensor]:
+        """Evaluate the expression for multiple complex inputs and return the result."""
+
+class Representation:
+    r"""
+    A representation class in the sense of representation theory. This class is used to represent the representation of a tensor. It is essentially a pair of a name and a dimension.
+    New representations are registered when constructing.
+    Some representations are dualizable, meaning that they have a dual representation.
+    Indices will only ever match across dual representations.
+    There are some already registered representations, such as:
+     EUCLIDEAN: Rep = Rep::SelfDual(0);
+     BISPINOR: Rep = Rep::SelfDual(1);
+     COLORADJ: Rep = Rep::SelfDual(2);
+     MINKOWSKI: Rep = Rep::SelfDual(3);
+
+     LORENTZ_UP: Rep = Rep::Dualizable(1);
+     LORENTZ_DOWN: Rep = Rep::Dualizable(-1);
+     SPINFUND: Rep = Rep::Dualizable(2);
+     SPINANTIFUND: Rep = Rep::Dualizable(-2);
+     COLORFUND: Rep = Rep::Dualizable(3);
+     COLORANTIFUND: Rep = Rep::Dualizable(-3);
+     COLORSEXT: Rep = Rep::Dualizable(4);
+     COLORANTISEXT: Rep = Rep::Dualizable(-4);
+    """
+    def __new__(cls,name:str,dimension:int,dual:bool =False): ...
+    def __call__(self, aind:int|Expression|str) -> Slot:
+        r"""
+        Generate a new slot with the given index, from this representation
+        """
+        ...
+
+    def __repr__(self) -> str:
+        ...
+
+    def __str__(self) -> str:
+        ...
+
+    def to_expression(self) -> Expression:
+        r"""
+        Return the symbolica expression of the representation.
+        """
+
+class Slot:
+    r"""
+    An abstract index slot for a tensor.
+    This is essentially a tuple of a `Representation` and an abstract index id.
+
+    The abstract index id can be either an integer or a symbol.
+    This is the building block for creating tensor structures that can be contracted.
+    """
+    def __new__(cls,name:str,dimension:int,aind:int|str|Expression,dual:bool =False): ...
+    def __repr__(self) -> str:
+        ...
+
+    def __str__(self) -> str:
+        ...
+
+    def to_expression(self) -> Expression:
+        r"""
+        Return the symbolica expression of the slot.
+        """
+
+
+class Tensor(Sequence):
+    r"""
+    A tensor class that can be either dense or sparse.
+    The data is either float or complex or a symbolica expression
+    It can be instantiated with data using the `sparse_empty` or `dense` module functions.
+    """
+    def structure(self) -> TensorStructure|TensorIndices:
+        r"""
+        Return the structure of the tensor.
+        """
+
+    def to_dense(self):
```

Comment:
I believe VS code gives better hints if the function returns the type `-> None` explicitly, now it yields `Any`.

Reply:
Alright will do!